File: lib/domain/models/mentsu.dart
import '../../utils/tile_parser.dart';

enum MentsuType {
  shuntsu, // 順子
  minko,   // 明刻
  anko,    // 暗刻
  minkan,  // 明槓
  ankan,   // 暗槓
}

class Mentsu {
  final MentsuType type;
  final List<Tile> tiles;

  const Mentsu({required this.type, required this.tiles});

  bool get isShuntsu => type == MentsuType.shuntsu;
  bool get isKotsu => type == MentsuType.minko || type == MentsuType.anko;
  bool get isKantsu => type == MentsuType.minkan || type == MentsuType.ankan;
  bool get isOpen => type == MentsuType.minko || type == MentsuType.minkan;
  bool get isClosed => !isOpen;

  bool get containsTerminalOrHonor {
    return tiles.any((t) =>
        t.type == TileType.wind ||
        t.type == TileType.dragon ||
        t.number == 1 ||
        t.number == 9);
  }

  int get fuValue {
    if (isShuntsu) return 0;
    final isYaochu = containsTerminalOrHonor;
    switch (type) {
      case MentsuType.minko:
        return isYaochu ? 4 : 2;
      case MentsuType.anko:
        return isYaochu ? 8 : 4;
      case MentsuType.minkan:
        return isYaochu ? 16 : 8;
      case MentsuType.ankan:
        return isYaochu ? 32 : 16;
      default:
        return 0;
    }
  }
}

File: src/domain/models/mentsu.ts
import { Tile, TileType } from '../../utils/tileParser';

export enum MentsuType {
  shuntsu = 'shuntsu', // 順子
  minko = 'minko',     // 明刻
  anko = 'anko',       // 暗刻
  minkan = 'minkan',   // 明槓
  ankan = 'ankan',     // 暗槓
}

export class Mentsu {
  readonly type: MentsuType;
  readonly tiles: Tile[];

  constructor(type: MentsuType, tiles: Tile[]) {
    this.type = type;
    this.tiles = tiles;
  }

  get isShuntsu(): boolean { return this.type === MentsuType.shuntsu; }
  get isKotsu(): boolean { return this.type === MentsuType.minko || this.type === MentsuType.anko; }
  get isKantsu(): boolean { return this.type === MentsuType.minkan || this.type === MentsuType.ankan; }
  get isOpen(): boolean { return this.type === MentsuType.minko || this.type === MentsuType.minkan; }
  get isClosed(): boolean { return !this.isOpen; }

  get containsTerminalOrHonor(): boolean {
    return this.tiles.some(t =>
      t.type === TileType.wind ||
      t.type === TileType.dragon ||
      t.number === 1 ||
      t.number === 9
    );
  }

  get fuValue(): number {
    if (this.isShuntsu) return 0;
    const isYaochu = this.containsTerminalOrHonor;
    switch (this.type) {
      case MentsuType.minko:  return isYaochu ? 4 : 2;
      case MentsuType.anko:   return isYaochu ? 8 : 4;
      case MentsuType.minkan: return isYaochu ? 16 : 8;
      case MentsuType.ankan:  return isYaochu ? 32 : 16;
      default: return 0;
    }
  }
}
